### PR TITLE
Bluetooth: Gatt: Automatic discover of CCC when subscribing

### PR DESCRIPTION
**Pull request by Emil-Gydesen-Bose**
Added a feature that allows an application to request
automatic discovery of the CCCD handle when subscriping
to a characteristic. In order to preserver RAM, the application
is expected to provide the discovery parameters, and it's up
to the application whether or not it wants to reuse the
discovery parameters or use one for each characteristic.

Signed-off-by: Emil Gydesen <emil_gydesen@bose.com>

### CODE REVIEW

**Comment by Emil-Gydesen-Bose** [comment]
I would have liked to avoid putting the `struct bt_gatt_subscribe_params *sub_params;` in the `struct bt_gatt_discover_params`, but I couldn't figure out any other way to get the `bt_gatt_subscribe_params` in `gatt_ccc_discover_cb`, except if the `bt_gatt_subscribe_params` contained the `bt_gatt_discover_params` rather than just having the pointer.

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -3929,6 +3966,30 @@ int bt_gatt_subscribe(struct bt_conn *conn,
 	if (!has_subscription) {
 		int err;
 
+#if defined(CONFIG_BT_GATT_AUTO_DISCOVER_CCC)
```
Comment (Vudentz) [inline subsys/bluetooth/host/gatt.c:3969]: Id move this block to its own function.

**Review by Vudentz — APPROVED** [review]
Other than that it looks pretty good to me.

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (Emil-Gydesen-Bose) [reply subsys/bluetooth/host/gatt.c:3969]: Good idea

File: subsys/bluetooth/host/Kconfig.gatt [hunk 2]
```diff
@@ -138,6 +138,14 @@ config BT_GATT_READ_MULTIPLE
 	  This option enables support for the GATT Read Multiple Characteristic
 	  Values procedure.
 
+config BT_GATT_AUTO_DISCOVER_CCC
+	bool "Support to automatic discover the CCC handles of characteristics"
+	default n
```
Comment (joerchan) [inline subsys/bluetooth/host/Kconfig.gatt:143]: default is 'n', should be removed. (somebody went through the entire code-base and cleaned this up earlier).

File: include/bluetooth/gatt.h [hunk 3]
```diff
@@ -1145,6 +1145,10 @@ struct bt_gatt_discover_params {
 	uint16_t end_handle;
 	/** Discover type */
 	uint8_t type;
+#if defined(CONFIG_BT_GATT_AUTO_DISCOVER_CCC)
+	/** The sub_params, if any, used for automatic discover */
```
Comment (joerchan) [inline include/bluetooth/gatt.h:1149]: This parameter shouldn't be set by the application, right?
It should be documented.

`/** Only for stack-internal use, used for automatic discovery. */`

**Review by joerchan — CHANGES_REQUESTED** [review]
Implementation looks good.

File: subsys/bluetooth/host/Kconfig.gatt (hunk 2 above)
Comment (Emil-Gydesen-Bose) [reply subsys/bluetooth/host/Kconfig.gatt:143]: Roger that, removed.

File: include/bluetooth/gatt.h (hunk 3 above)
Comment (Emil-Gydesen-Bose) [reply include/bluetooth/gatt.h:1149]: Correct. Updated the comment.

**Review by joerchan — APPROVED** [review]

**Pull request merged**